src/traces/box/defaults.js

```diff
@@ -22,49 +22,183 @@ function supplyDefaults(traceIn, traceOut, defaultColor, layout) {
     if(notched) coerce('notchwidth');
 
     handlePointsDefaults(traceIn, traceOut, coerce, {prefix: 'box'});
 }
 
 function handleSampleDefaults(traceIn, traceOut, coerce, layout) {
+    function getDims(arr) {
+        var dims = 0;
+        if(arr && arr.length) {
+            dims += 1;
+            if(Lib.isArrayOrTypedArray(arr[0]) && arr[0].length) {
+                dims += 1;
+            }
+        }
+        return dims;
+    }
+
+    function valid(astr) {
+        return Lib.validate(traceIn[astr], attributes[astr]);
+    }
+
     var y = coerce('y');
     var x = coerce('x');
-    var hasX = x && x.length;
+
+    var sLen;
+    if(traceOut.type === 'box') {
+        var q1 = coerce('q1');
+        var median = coerce('median');
+        var q3 = coerce('q3');
+
+        traceOut._hasPreCompStats = (
+            q1 && q1.length &&
+            median && median.length &&
+            q3 && q3.length
+        );
+        sLen = Math.min(
+            Lib.minRowLength(q1),
+            Lib.minRowLength(median),
+            Lib.minRowLength(q3)
+        );
+    }
+
+    var yDims = getDims(y);
+    var xDims = getDims(x);
+    var yLen = yDims && Lib.minRowLength(y);
+    var xLen = xDims && Lib.minRowLength(x);
 
     var defaultOrientation, len;
+    if(traceOut._hasPreCompStats) {
+        switch(String(xDims) + String(yDims)) {
+            // no x / no y
+            case '00':
+                var setInX = valid('x0') || valid('dx');
+                var setInY = valid('y0') || valid('dy');
+
+                if(setInY && !setInX) {
+                    defaultOrientation = 'h';
+                } else {
+                    defaultOrientation = 'v';
+                }
 
-    if(y && y.length) {
+                len = sLen;
+                break;
+            // just x
+            case '10':
+                defaultOrientation = 'v';
+                len = Math.min(sLen, xLen);
+                break;
+            case '20':
+                defaultOrientation = 'h';
+                len = Math.min(sLen, xLen);
+                break;
+            // just y
+            case '01':
+                defaultOrientation = 'h';
+                len = Math.min(sLen, yLen);
+                break;
+            case '02':
+                defaultOrientation = 'v';
+                len = Math.min(sLen, yLen);
+                break;
+            // both
+            case '12':
+                defaultOrientation = 'v';
+                len = Math.min(sLen, xLen, yLen);
+                break;
+            case '21':
+                defaultOrientation = 'h';
+                len = Math.min(sLen, xLen, yLen);
+                break;
+            case '11':
+                // this one is ill-defined
+                len = 0;
+                break;
+            case '22':
+                // this one case happen on multi-category axes
+                defaultOrientation = 'v';
+                len = Math.min(sLen, xLen, yLen);
```

Comment:
Doesn't the result here depend on which axis is multi-category?

Reply:
Yeah, good call. The array on the multicategory axis would corresponding to positions (by default).

Reply:
done in https://github.com/plotly/plotly.js/pull/4432/commits/7ecf81ddacbd1e79b108ac7458225980c308f741